src/libraries/Microsoft.Extensions.Options/src/OptionsCache.cs

```diff
@@ -31,14 +31,44 @@ public class OptionsCache<[DynamicallyAccessedMembers(Options.DynamicallyAccesse
         public virtual TOptions GetOrAdd(string? name, Func<TOptions> createOptions!!)
         {
             name ??= Options.DefaultName;
-            Lazy<TOptions>? value;
+            Lazy<TOptions> value;
 
-#if NETSTANDARD2_1
+#if NET || NETSTANDARD2_1
             value = _cache.GetOrAdd(name, static (name, createOptions) => new Lazy<TOptions>(createOptions), createOptions);
 #else
             if (!_cache.TryGetValue(name, out value))
             {
-                value = _cache.GetOrAdd(name, new Lazy<TOptions>(createOptions));
+                Func<TOptions> localCreateOptions = createOptions;
+                value = _cache.GetOrAdd(name, new Lazy<TOptions>(localCreateOptions));
+            }
+#endif
+
+            return value.Value;
+        }
+
+        internal TOptions GetOrAdd<TArg>(string? name, Func<string, TArg, TOptions> createOptions, TArg factoryArgument)
+        {
+            // For compatibility, fall back to public GetOrAdd() if we're in a derived class.
+            if (GetType() != typeof(OptionsCache<TOptions>))
+            {
+                string? localName = name;
+                Func<string, TArg, TOptions> localCreateOptions = createOptions;
+                TArg localFactoryArgument = factoryArgument;
+                return GetOrAdd(name, () => localCreateOptions(localName ?? Options.DefaultName, localFactoryArgument));
```

Comment:
Could you explain in a comment here (and at similar places in your change) why you store the value in a local for this.

Then the code is easier to understand :)